### ISSUE TITLE
[Feature Request] Rebindable motion input controls to mouse/keyboard/other controllers

### ISSUE THREAD

**Issue by ElTioRata** [RPCS3/rpcs3#13883]
### Quick summary

Bakugan (and other games) requires you to have a PS3 controller with motion input because various game mechanics are tied to it. Unfortunately, at least for the case of Bakugan, the game doesn't provide a "non-motion" controller option, so the game is unplayable on PC controls and standard controllers. A workaround for this issue could be making the motion input fully rebindable, a good example of this feature is Dolphin with its Wiimote.

**Comment by AIGLE25** [comment]
yeah sure ! i am stuck on sly cooper too

**Comment by KonstantinosK77** [comment]
I thin it's necessary!!
A lot of us we have Xinput game controllers and we can't play games with motion control!!

**Comment by adam-sunderman** [comment]
I would like this as well! I wanna play Thieves in Time :)

**Comment by Valtekken** [comment]
+1 on this, Killzone 2 is unplayable otherwise and my DS3 is borked so I don't even have the option

**Comment by JimScript** [comment]
I wonder how this could be implemented, Sixaxis _seems_ to actually be determined by 4 values(or at least that's what RPCS3 reads for a Dualsense):

- The X Axis: how it tilts left(-) or right(+)
- The Y Axis: how right-side-up(-) or upside-down(+) it is
- The Z Axis: how it tilts forward(-) or backward(+)
- The G Axis: how fast it is moving counterclockwise(-) or clockwise(+)

The range of values seem to be centered around 512 with +/- ~112 at rest and the numbers going wildly if you shake the controller. For all those tilting mini-games, you could map the X and Z to a stick, but I'm not sure if emulating the Y axis would be necessary to add. For all those shaking inputs, bouncing the G axis around might be enough to trigger it, but every developer could read "shake" a different way(unless there's a hidden system call that just tells the game a controller is shaking).

**Comment by adam-sunderman** [comment]
Yeah I think the 4 values means you would need to either A target common use cases or B allow a bit of mapping and custom functionality. I think mapping x and y to the stick would serve a vast percentage of use cases.

**Comment by gsergeantGTHB** [comment]
+1

**Comment by luphoria** [comment]
I've made a post to the RPCS3 forums regarding this feature: https://forums.rpcs3.net/showthread.php?tid=208083 The forum post still needs to be manually approved. 

I'm interested in perhaps tackling this issue but I've never contributed to RPCS3 before. 

From my post:
> In my searching for answers, I also came across this repo: https://github.com/ZeptoBST/DualShock4-emulator/ -- which appears able to emulate SIXAXIS from xbox and keyboard controller inputs, very much like what I want. I don't know if the tool actually features rebindable keys for this or a joystick option, though, because/and it's Windows-only, and I'm on Linux. If there is some kind of solution for this that's cross-platform and works on generally any gamepad, that would suit my needs. 
> It seems like under the hood there is already SIXAXIS emulation to some degree, because I was able to get gyro working from my Switch Pro controller without any hiccups. This implies that there is likely some way I can plug in other inputs wherever the sdl gyro is being translated to ds3 sixaxis. If it's possible for me to implement this feature myself, I would love to be able to contribute upstream as well. But I don't really know how to get started in the rpcs3 repository.

**Comment by luphoria** [comment]
https://github.com/RPCS3/rpcs3/blob/71524271e948316d57515422bd0da0159a55d24d/rpcs3/Input/sdl_pad_handler.cpp#L718

It should be pretty simple to take the gyro -> sixaxis approach like this to stick -> sixaxis. Still kind of wandering in my understanding the codebase here, but it shouldn't be too difficult. Should just be math with existing numbers to convert the current stick position to DS3 resolution format for yaw.

**Comment by luphoria** [comment]
https://github.com/RPCS3/rpcs3/assets/60309933/9494e7f7-e279-4c59-afb6-0a38586f3946

I've got gyro emulation on left stick evdev working! As you can see in the video, it's very jank. I don't know what normal values look like (other than that they are between 0 and 1024) and I'm not a mathematician so I just did a very basic conversion from the joystick range to the gyro range. 

I added the following lines after rpcs3/Input/evdev_joystick_handler.cpp:L1221: 
```cpp
pad->m_sensors[0].m_value = Clamp0To1023(512 + (lx - 127.5f) * 4.016f);
pad->m_sensors[1].m_value = Clamp0To1023(512 + (ly - 127.5f) * 4.016f); 
```
Just looking at it, my equation for what i'm doing is overcomplicated lol. But regardless it's just a successful test. I'm gonna rrefine the conversion so it's smoother. 
If you are feeling spicy, you could even use the right stick for Z and W values. 

If there's interest, I would be happy to try making this an actual feature instead of a hacky thing.

**Comment by luphoria** [comment]
Not a solution to good equation, but this code makes the input feel pretty natural (at least for the game I'm playing): 
```cpp

	// apply deadzone (is the configured deadzone accessible here?)
	if (abs(lx - 127) < 30)
		lx = 127.5f;
	if (abs(ly - 127) < 30)
		ly = 127.5f;

	pad->m_sensors[0].m_value = Clamp0To1023(512 + (lx - 127.5f));
	pad->m_sensors[1].m_value = Clamp0To1023(399 + (ly - 127.5f)); 
```

**Comment by luphoria** [comment]
Moved the logic to rpcs3/Input/pad_thread.cpp after line 538:
```cpp

		for (u32 i = 0; i < connected_devices; i++)
		{
			const auto& pad = m_pads[0];
			pad->m_sensors[0].m_value = (512 + (pad->m_sticks[0].m_value - 127.5f));
			pad->m_sensors[1].m_value = (399 + (pad->m_sticks[1].m_value - 127.5f) * 2);
		}
```
This should bind gyro X and Y to the left joystick for all input types.

**Comment by luphoria** [comment]
Branch with this change: https://github.com/luphoria/rpcs3/tree/joystick-rebind
I don't know if I plan on ever making this upstream-ready, but I do intend to flesh out the feature.

**Comment by sYnapZiX** [comment]
Just allow the inputs to be mapped onto the mouse (which allows for analog input) or even the keyboard (binary controls only) and everything will be fine. Other emulators like Cemu, Yuzu (Suyu), Ryujinz have that feature without the need for additional tinkering outside the emulator.

**Comment by LeBlackMeow** [comment]
Still no workaround for SIXAXIS emulation ?

**Comment by HappehLemons** [comment]
No progress on this one? Unless using a motion controller this limits a lot of games.

**Comment by Gammler** [comment]
without this you can't use tilt in guitar hero, because guitar hero uses sixaxis to detect starpower activation when tilting the guitar.

**Comment by Xbox-Series-X-User** [comment]
> without this you can't use tilt in guitar hero, because guitar hero uses sixaxis to detect starpower activation when tilting the guitar.

Tilt? I always used the select button

**Comment by AniLeo** [comment]
It won't implement itself now nor in 2077... Be the change you want

**Comment by GHusr-y5** [comment]
> so here's a Windows binary with SIXAXIS's X and Y value mapped to the left joystick. 😄

Left joystick is perfect, thanks! I has no skills in doing something more complicated thaт complaining on forums, so already completed build is a miracle! Thanks a lot!

**Comment by luphoria** [comment]
Video tutorial & explanation: https://www.youtube.com/watch?v=rnbu9Gueiyc

**Comment by luphoria** [comment]
This video above has gotten 3.7K views on YouTube in just a few months; I think this is pretty clearly a highly-requested feature. Obviously my implementation is not upstream ready; I took a look at how a proper implementation might look, and it seems like a lot more work than I'm cut out for. 

tldr... If you are somebody with more experience and free time than me, and you are looking for a feature that lots of people would benefit from, I think that this particular issue might be a good focus for you. :-)

**Comment by madman-asunder** [comment]
thanks @luphoria 
I was able to use what you learned to make a build where the left stick is bound to X and Z which works really well for sly cooper thieves in time.

Here is my build. please watch luphoria's youtube video for instructions explanation
tutorial:
https://www.youtube.com/watch?v=rnbu9Gueiyc

rpcs3-gyro-LS-xz
https://drive.google.com/file/d/1WTQtz7aKG_hT_pbUdJQs7lW9o1qjk8su/view?usp=sharing

**Comment by Windsurf7** [comment]
This issue is now half-resolved via #18113.
Use the "toggle mouse-based gyro" hotkey and the mouse while holding the right mouse button, in order to emulate gyro motion controls.

**Comment by ElTioRata** [comment]
Finally, can confirm that Killzone 2 first level is now passable. ~Will test Bakugan next.~ Doesn't seem to work with Bakugan (you have to tilt the controller in this case). Would also be nice to have an icon which tells you if the toggle is enabled.

**Comment by Windsurf7** [comment]
Have you tried all axes?
Mouse position controls gyro X/Z, mouse wheel controls gyro Y.
X/Z is the top-down view, confirmed by Sly 4's Spark Runner minigame.

**Comment by ElTioRata** [comment]
Yes I tried all axes (left, right, up, down and scroll).

**Comment by Windsurf7** [comment]
Hmm I don't know what to say, perhaps it uses the G axis, which I didn't implement. It seems to be an additional axis that implies how much force is put into the speed. If anyone ever implements this to be rebindable in the pad configuration, that should be included.
